qualtran/bloqs/gf_arithmetic/gf2_multiplication.py

```diff
@@ -653,3 +660,333 @@ def _binarypolynomialmultiplication() -> BinaryPolynomialMultiplication:
+class GF2ShiftRightMod(Bloq):
+    r"""Multiplies by $2^k$ (or $x^k$ for polynomials) modulu.
+
+    Applies the transformation
+    $$
+        \ket{f} \rightarrow \ket{x^k f \mod m(x)}
+    $$
+
+    Where the modulus $m(x)$ is the irreducible polynomial defining the galois field arithmetic.
+
+    Args:
+        m_x: The irreducible polynomial that defines the galois field.
+        k: The number of shifts (i.e. the exponent of $2$ or $x$).
+
+    Registers:
+        f: The number (polynomial) to shift.
+
+    References:
+        [Space-efficient quantum multiplication of polynomials for binary finite fields with
+            sub-quadratic Toffoli gate count](https://arxiv.org/abs/1910.02849v2) Section 3.1
+    """
+
+    m_x: Poly = attrs.field(converter=lambda x: x if isinstance(x, Poly) else Poly.Degrees(x))
+    k: SymbolicInt = 1
+
+    @cached_property
+    def n(self):
+        return self.m_x.degree
+
+    @cached_property
+    def qgf(self):
+        return QGF(2, self.n, irreducible_poly=self.m_x)
+
+    @cached_property
+    def gf(self):
+        return GF(2, self.n, irreducible_poly=self.m_x if self.n > 2 else None)
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        return Signature([Register('f', dtype=self.qgf)])
+
+    @cached_property
+    def degrees(self):
+        return tuple(sorted(self.m_x.nonzero_degrees))
+
+    def on_classical_vals(self, f: 'ClassicalValT') -> Dict[str, 'ClassicalValT']:
+        k = self.k
+        if is_symbolic(k):
+            raise TypeError(f'classical action is not supported for {self}')
+        assert isinstance(f, self.gf)
+        if k == 0 or self.n == 1:
+            return {'f': f}
+        return {'f': f * self.gf(2) ** k}
+
+    def build_composite_bloq(self, bb: 'BloqBuilder', f: 'Soquet') -> Dict[str, 'SoquetT']:
+        if is_symbolic(self.k):
+            raise DecomposeTypeError(f'symbolic decomposition is not supported for {self}')
+        f_arr = bb.split(f)[::-1]
+        if self.n > 1:
+            for _ in range(self.k):
+                f_arr = np.roll(f_arr, 1)
+                for i in self.degrees[1:-1]:
+                    f_arr[0], f_arr[i] = bb.add(CNOT(), ctrl=f_arr[0], target=f_arr[i])
+            f_arr = f_arr[::-1]
+        f = bb.join(f_arr)
+        return {'f': f}
+
+    def build_call_graph(
+        self, ssa: 'SympySymbolAllocator'
+    ) -> Union['BloqCountDictT', Set['BloqCountT']]:
+        if self.k == 0 or self.n == 1:
+            return {}
+        return {CNOT(): max(len(self.degrees) - 2, 0) * self.k}
+
+
+@bloq_example
+def _gf2shiftrightmod() -> GF2ShiftRightMod:
+    m_x = [5, 2, 0]  # x^5 + x^2 + 1
+    gf2shiftrightmod = GF2ShiftRightMod(m_x=m_x, k=3)  # shift by 3
+    return gf2shiftrightmod
+
+
+_GF2_SHIFT_RIGHT_MOD_DOC = BloqDocSpec(bloq_cls=GF2ShiftRightMod, examples=(_gf2shiftrightmod,))
+
+
+@attrs.frozen
+class _GF2MulImpl(Bloq):
+    """Multiply two GF2 numbers (or binary polynomials) modulu m(x)."""
```

Comment:
Please add a more detailed docstring if possible. How is this bloq different from `GF2Mul` below ?

Reply:
done